File: dotnet/Falcon.PluginAPI/Services/Services.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class Services
    {
        [ThreadStatic]
        private static nint _response;

        private readonly nint _plugin;

        internal Services(nint plugin)
        {
            _plugin = plugin;
        }

        public bool Provide(string name, Func<string, string> handler)
        {
            ArgumentNullException.ThrowIfNull(handler);
            void* userData = PluginCallback.Keep(_plugin, handler);
            bool provided;
            fixed (byte* text = NativeApi.Utf8(name))
            {
                provided = NativeApi.Table->registerService((FalconPlugin*)_plugin, text, &Serve, userData) != 0;
            }

            if (!provided)
            {
                PluginCallback.Release(userData);
            }

            return provided;
        }

        public void Withdraw(string name)
        {
            fixed (byte* text = NativeApi.Utf8(name))
            {
                NativeApi.Table->unregisterService((FalconPlugin*)_plugin, text);
            }
        }

        public bool Has(string name)
        {
            fixed (byte* text = NativeApi.Utf8(name))
            {
                return NativeApi.Table->hasService(text) != 0;
            }
        }

        public string? Provider(string name)
        {
            fixed (byte* text = NativeApi.Utf8(name))
            {
                byte* plugin = NativeApi.Table->serviceProvider(text);
                if (plugin == null)
                {
                    return null;
                }

                return NativeApi.Text(plugin);
            }
        }

        public string? Call(string name, string request = "")
        {
            int found = 0;
            byte* response;
            fixed (byte* nameText = NativeApi.Utf8(name))
            fixed (byte* requestText = NativeApi.Utf8(request))
            {
                response = NativeApi.Table->callService(nameText, requestText, &found);
            }

            if (found == 0)
            {
                return null;
            }

            return NativeApi.Text(response);
        }

        public ulong On(string name, Action<CustomEvent> handler, EventPriority priority = EventPriority.Normal,
                        bool ignoreCancelled = false)
        {
            ArgumentNullException.ThrowIfNull(handler);
            Action<nint> dispatch = handle =>
            {
                handler(CustomEvent.Create(handle));
            };

            void* userData = PluginCallback.Keep(_plugin, dispatch);
            ulong subscription;
            fixed (byte* text = NativeApi.Utf8(name))
            {
                subscription = NativeApi.Table->subscribeCustomEvent((FalconPlugin*)_plugin, text, (uint)priority,
                                                                     ignoreCancelled ? 1 : 0, &Events.Dispatch,
                                                                     userData);
            }

            if (subscription == 0)
            {
                PluginCallback.Release(userData);
            }

            return subscription;
        }

        public CustomEventResult Fire(string name, string data = "", bool cancellable = false)
        {
            int cancelled = 0;
            byte* result;
            fixed (byte* nameText = NativeApi.Utf8(name))
            fixed (byte* dataText = NativeApi.Utf8(data))
            {
                result = NativeApi.Table->fireCustomEvent((FalconPlugin*)_plugin, nameText, dataText,
                                                          cancellable ? 1 : 0, &cancelled);
            }

            return new CustomEventResult(cancelled != 0, NativeApi.Text(result));
        }

        [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
        private static byte* Serve(byte* request, void* userData)
        {
            string answer = PluginCallback.Run(userData, false, "A service handler threw an exception", (nint)request,
                                               static (target, state) =>
                                               {
                                                   string text = NativeApi.Text((byte*)state);
                                                   return ((Func<string, string>)target)(text) ?? string.Empty;
                                               }, string.Empty);
            return Hold(answer);
        }

        private static byte* Hold(string value)
        {
            if (_response != 0)
            {
                Marshal.FreeCoTaskMem(_response);
                _response = 0;
            }

            _response = Marshal.StringToCoTaskMemUTF8(value);
            return (byte*)_response;
        }
    }
}

File: java/src/main/java/falcon/api/service/Services.java
package falcon.api.service;

import falcon.api.event.CustomEvent;
import falcon.api.event.EventPriority;
import falcon.api.internal.FalconAbi;
import falcon.api.internal.Interop;
import falcon.api.internal.PluginContext;
import falcon.api.internal.Upcalls;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.lang.invoke.MethodHandles;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

public final class Services {
    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
    private static final MemorySegment SERVE = Upcalls.stub(LOOKUP, "serve", FalconAbi.SERVICE_HANDLER);
    private static final MemorySegment DISPATCH = Upcalls.stub(LOOKUP, "dispatch", FalconAbi.EVENT_HANDLER);
    private static final ThreadLocal<Arena> RESPONSES = new ThreadLocal<>();

    private final PluginContext mContext;
    private final Map<String, MemorySegment> mProvided = new ConcurrentHashMap<>();

    public Services(PluginContext context) {
        mContext = context;
    }

    public boolean provide(String name, ServiceHandler handler) {
        Objects.requireNonNull(handler, "handler");
        MemorySegment userData = Upcalls.register(new Provider(mContext, handler));
        try (Arena arena = Arena.ofConfined()) {
            boolean provided = Interop.api().registerService(mContext.handle(), Interop.text(arena, name), SERVE,
                    userData) != 0;
            if (!provided) {
                Upcalls.release(userData, Provider.class);
                return false;
            }
        }

        MemorySegment previous = mProvided.put(name, userData);
        if (previous != null) {
            Upcalls.release(previous, Provider.class);
        }
        return true;
    }

    public void withdraw(String name) {
        try (Arena arena = Arena.ofConfined()) {
            Interop.api().unregisterService(mContext.handle(), Interop.text(arena, name));
        }
        MemorySegment userData = mProvided.remove(name);
        if (userData != null) {
            Upcalls.release(userData, Provider.class);
        }
    }

    public boolean has(String name) {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.api().hasService(Interop.text(arena, name)) != 0;
        }
    }

    public Optional<String> provider(String name) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment plugin = Interop.api().serviceProvider(Interop.text(arena, name));
            if (Interop.isNull(plugin)) {
                return Optional.empty();
            }
            return Optional.of(Interop.string(plugin));
        }
    }

    public Optional<String> call(String name) {
        return call(name, "");
    }

    public Optional<String> call(String name, String request) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment found = arena.allocate(ValueLayout.JAVA_INT);
            MemorySegment response = Interop.api().callService(Interop.text(arena, name),
                    Interop.text(arena, request), found);
            if (found.get(ValueLayout.JAVA_INT, 0) == 0) {
                return Optional.empty();
            }
            return Optional.of(Interop.string(response));
        }
    }

    public long on(String name, Consumer<? super CustomEvent> handler) {
        return on(name, handler, EventPriority.NORMAL, false);
    }

    public long on(String name, Consumer<? super CustomEvent> handler, EventPriority priority) {
        return on(name, handler, priority, false);
    }

    public long on(String name, Consumer<? super CustomEvent> handler, EventPriority priority,
                   boolean ignoreCancelled) {
        Objects.requireNonNull(handler, "handler");
        MemorySegment userData = Upcalls.register(new Listener(mContext, handler));
        try (Arena arena = Arena.ofConfined()) {
            long id = Interop.api().subscribeCustomEvent(mContext.handle(), Interop.text(arena, name),
                    priority.value(), ignoreCancelled ? 1 : 0, DISPATCH, userData);
            if (id == 0) {
                Upcalls.release(userData, Listener.class);
            }
            return id;
        }
    }

    public CustomEventResult fire(String name) {
        return fire(name, "", false);
    }

    public CustomEventResult fire(String name, String data) {
        return fire(name, data, false);
    }

    public CustomEventResult fire(String name, String data, boolean cancellable) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment cancelled = arena.allocate(ValueLayout.JAVA_INT);
            MemorySegment result = Interop.api().fireCustomEvent(mContext.handle(), Interop.text(arena, name),
                    Interop.text(arena, data), cancellable ? 1 : 0, cancelled);
            return new CustomEventResult(cancelled.get(ValueLayout.JAVA_INT, 0) != 0, Interop.string(result));
        }
    }

    private static MemorySegment serve(MemorySegment request, MemorySegment userData) {
        try {
            Provider provider = Upcalls.target(userData, Provider.class);
            String text = Interop.string(request);
            String[] response = {""};
            if (provider != null) {
                provider.context().guard("A service handler threw an exception", () -> {
                    response[0] = provider.handler().handle(text);
                });
            }
            return respond(response[0]);
        } catch (Throwable ignored) {
            return MemorySegment.NULL;
        }
    }

    private static MemorySegment respond(String response) {
        Arena previous = RESPONSES.get();
        if (previous != null) {
            previous.close();
        }
        Arena arena = Arena.ofConfined();
        RESPONSES.set(arena);
        return Interop.text(arena, response);
    }

    private static void dispatch(MemorySegment event, MemorySegment userData) {
        try {
            Listener listener = Upcalls.target(userData, Listener.class);
            if (listener != null) {
                CustomEvent custom = new CustomEvent(listener.context(), event);
                listener.context().guard("A custom event handler threw an exception",
                        () -> listener.handler().accept(custom));
            }
        } catch (Throwable ignored) {
        }
    }

    private record Provider(PluginContext context, ServiceHandler handler) {
    }

    private record Listener(PluginContext context, Consumer<? super CustomEvent> handler) {
    }
}
